# Reglas de asociación

In [1]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

In [2]:
data = pd.read_csv('../Datasets/GroceryStoreDataSet.csv',names = ['products'],sep = ',')
data

products
0               MILK,BREAD,BISCUIT
1    BREAD,MILK,BISCUIT,CORNFLAKES
2              BREAD,TEA,BOURNVITA
3             JAM,MAGGI,BREAD,MILK
4                MAGGI,TEA,BISCUIT
5              BREAD,TEA,BOURNVITA
6             MAGGI,TEA,CORNFLAKES
7          MAGGI,BREAD,TEA,BISCUIT
8              JAM,MAGGI,BREAD,TEA
9                       BREAD,MILK
10  COFFEE,COCK,BISCUIT,CORNFLAKES
11  COFFEE,COCK,BISCUIT,CORNFLAKES
12          COFFEE,SUGER,BOURNVITA
13               BREAD,COFFEE,COCK
14             BREAD,SUGER,BISCUIT
15         COFFEE,SUGER,CORNFLAKES
16           BREAD,SUGER,BOURNVITA
17              BREAD,COFFEE,SUGER
18              BREAD,COFFEE,SUGER
19      TEA,MILK,COFFEE,CORNFLAKES

In [3]:
data.shape

(20, 1)

Dividimos los productos y creamos una lista llamada dataf

In [4]:
dataf = list(data["products"].apply(lambda x:x.split(",") ))
dataf

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

Transformamos la lista en un dataframe de 0 y 1

In [5]:
a = TransactionEncoder()
a_dataf = a.fit(dataf).transform(dataf)
data = pd.DataFrame(a_dataf,columns=a.columns_)
data = data.replace(False,0)
data

BISCUIT BOURNVITA BREAD  COCK COFFEE CORNFLAKES   JAM MAGGI  MILK SUGER  \
0     True         0  True     0      0          0     0     0  True     0   
1     True         0  True     0      0       True     0     0  True     0   
2        0      True  True     0      0          0     0     0     0     0   
3        0         0  True     0      0          0  True  True  True     0   
4     True         0     0     0      0          0     0  True     0     0   
5        0      True  True     0      0          0     0     0     0     0   
6        0         0     0     0      0       True     0  True     0     0   
7     True         0  True     0      0          0     0  True     0     0   
8        0         0  True     0      0          0  True  True     0     0   
9        0         0  True     0      0          0     0     0  True     0   
10    True         0     0  True   True       True     0     0     0     0   
11    True         0     0  True   True       True     0     0     0     0   
12       0      True     0     0   True          0     0     0     0  True   
13       0         0  True  True   True          0     0     0     0     0   
14    True         0  True     0      0          0     0     0     0  True   
15       0         0     0     0   True       True     0     0     0  True   
16       0      True  True     0      0          0     0     0     0  True   
17       0         0  True     0   True          0     0     0     0  True   
18       0         0  True     0   True          0     0     0     0  True   
19       0         0     0     0   True       True     0     0  True     0   

     TEA  
0      0  
1      0  
2   True  
3      0  
4   True  
5   True  
6   True  
7   True  
8   True  
9      0  
10     0  
11     0  
12     0  
13     0  
14     0  
15     0  
16     0  
17     0  
18     0  
19  True

Frecuencia de los productos o items

In [6]:
data.sum()

BISCUIT        7
BOURNVITA      4
BREAD         13
COCK           3
COFFEE         8
CORNFLAKES     6
JAM            2
MAGGI          5
MILK           5
SUGER          6
TEA            7
dtype: object

Frecuenca de los productos o items

In [7]:
pd.DataFrame(data.sum() / data.shape[0], columns = ["Support"]).sort_values("Support", ascending = False)

Support
BREAD         0.65
COFFEE         0.4
BISCUIT       0.35
TEA           0.35
CORNFLAKES     0.3
SUGER          0.3
MAGGI         0.25
MILK          0.25
BOURNVITA      0.2
COCK          0.15
JAM            0.1

Aplicando el algoritmo apriori con soporte mínimo de 20%

In [8]:
df = apriori(data, min_support = 0.2, use_colnames = True)
df

c:\Users\User\AppData\Local\Programs\Python\Python311\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support              itemsets
0     0.35             (BISCUIT)
1      0.2           (BOURNVITA)
2     0.65               (BREAD)
3      0.4              (COFFEE)
4      0.3          (CORNFLAKES)
5     0.25               (MAGGI)
6     0.25                (MILK)
7      0.3               (SUGER)
8     0.35                 (TEA)
9      0.2      (BREAD, BISCUIT)
10     0.2         (BREAD, MILK)
11     0.2        (SUGER, BREAD)
12     0.2          (TEA, BREAD)
13     0.2  (COFFEE, CORNFLAKES)
14     0.2       (COFFEE, SUGER)
15     0.2          (TEA, MAGGI)

Usando una regla de asociación con una confianza mínima de 60%

In [9]:
df_ar = association_rules(df, metric = "confidence", min_threshold = 0.6)
df_ar

antecedents consequents  antecedent support  consequent support  support  \
0        (MILK)     (BREAD)                0.25                0.65      0.2   
1       (SUGER)     (BREAD)                0.30                0.65      0.2   
2  (CORNFLAKES)    (COFFEE)                0.30                0.40      0.2   
3       (SUGER)    (COFFEE)                0.30                0.40      0.2   
4       (MAGGI)       (TEA)                0.25                0.35      0.2   

   confidence      lift  leverage  conviction  zhangs_metric  
0    0.800000  1.230769    0.0375        1.75       0.250000  
1    0.666667  1.025641    0.0050        1.05       0.035714  
2    0.666667  1.666667    0.0800        1.80       0.571429  
3    0.666667  1.666667    0.0800        1.80       0.571429  
4    0.800000  2.285714    0.1125        3.25       0.750000

## A. Observaciones:
#### 1.La probabilidad de ver ventas de azúcar se estima en un 30%.
#### 2.La ingesta de pan se estima en un 65%.
#### 3. Podemos decir que el apoyo de ambos (azúcar y pan) se mide en un 20%.
#### 4. El 67% de los que compran azúcar, también compra pan.
#### 5. Los usuarios que compran azúcar probablemente consuman un 3% más de pan que los usuarios que no compran azúcar.
#### 6.Su correlación entre sí se considera 1.05.

## B. Sugerencias:
#### 1. La venta cruzada se puede mejorar combinando productos - artículos.
#### 2. El diseño del local se puede cambiar para que las ventas se puedan mejorar cuando ciertos artículos se mantienen juntos.
#### 3. Realizar actividades promocionales para incrementar las ventas de los productos que los clientes no compran.
#### 4. Ofrecer descuentos colectivos en estos productos si el cliente compra ambos.
